src/lib/components/TableOfContents/index.js

```diff
@@ -0,0 +1,154 @@
+/*
+ * Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {BaseElement} from '../BaseElement';
+
+/**
+ * Element that renders table of contents.
+ * @extends {BaseElement}
+ * @final
+ */
+class TableOfContents extends BaseElement {
+  static get properties() {
+    return {
+      title: {type: String},
+      titleId: {type: String, attribute: 'title-id'},
+    };
+  }
+
+  constructor() {
+    super();
+    this.close = this.close.bind(this);
+    this.open = this.open.bind(this);
+    this.scrollSpy = this.scrollSpy.bind(this);
+  }
+
+  connectedCallback() {
+    super.connectedCallback();
+
+    this.tocInnerDiv = document.createElement('div');
+    this.tocInnerDiv.classList.add('w-toc__content');
+    this.tocInnerDiv.innerHTML = this.innerText;
+    this.innerText = '';
+    this.append(this.tocInnerDiv);
+
+    this.divContent = document.querySelector('div#content');
+    this.headers = this.divContent.querySelectorAll(
+      'h1[id], h2[id], h3[id], h4[id], h5[id], h6[id]',
+    );
+
+    this.addTocDetails();
+    this.addOpenButton();
+
+    this.scrollSpy();
+    document.addEventListener('touchmove', this.scrollSpy);
+    document.addEventListener('scroll', this.scrollSpy);
+  }
+
+  disconnectedCallback() {
+    super.disconnectedCallback();
+    this.close();
+    document.removeEventListener('touchmove', this.scrollSpy);
+    document.removeEventListener('scroll', this.scrollSpy);
+  }
+
+  addOpenButton() {
+    const tocWrapper = document.createElement('div');
+    tocWrapper.classList.add('w-toc__button--wrapper');
+
+    const tocButton = document.createElement('button');
+    tocButton.classList.add(
+      'w-toc__button--open',
+      'w-button',
+      'w-button--secondary',
+      'w-button--icon',
+    );
+    tocButton.setAttribute('data-icon', 'list_alt');
+    tocButton.addEventListener('click', this.open);
+
+    tocWrapper.append(tocButton);
+    this.after(tocWrapper);
+  }
+
+  scrollSpy() {
```

Comment:
I noticed on this page https://deploy-preview-1983--web-dev-staging.netlify.app/prefers-color-scheme/
when you scroll near "Why dark mode" the active link disappears. I'm guessing because there's an h4 or something?